### PR TITLE
lock: Implement shared locking used for shared backend semantics.

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
This patch implements two things.

- Implements lock package implementing filesystem locks using
  FcntlFlock() syscall, also implements a compatible version
  on windows as well using win32 APIs.

- Adds a new ENV which indicates minio server to enable shared
  backend feature, while holding the locks on `format.json`

So two competing locks from two different minio processes compete
at filesystem level on a shared backend while performing an i/o.

This code is an attempt to allow this feature.

## Motivation and Context
Refer #3160 

## How Has This Been Tested?
make test and local shared backend setup.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by harshavardhana** [comment]
Do not need to review this just yet, there is more work to be done.

File: pkg/flock/lock.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
+package flock
+
+import (
+	"os"
+	"sync"
+)
+
+// FLocker - lock/unlock implementation backed by.
```
Comment (krisis) [inline pkg/flock/lock.go]: backed by a shared FS which implements fcntl(2).

File: pkg/flock/lock_test.go [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+	}
+	f.Close()
+	defer func() {
+		err = os.Remove(f.Name())
+		if err != nil {
+			t.Fatal(err)
+		}
+	}()
+
+	frw := NewFlock(f.Name())
```
Comment (krisis) [inline pkg/flock/lock_test.go]: Looks like the caller should create the file on which we hold lock. What should the caller do to hold a lock on a file yet to be created?

File: pkg/flock/lock.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+type lockedFile struct {
+	*os.File
+}
+
+// Private file mode.
+var privateFileMode = os.FileMode(0600)
+
+// RWLocker - interface that any read-write locking library should implement.
+type RWLocker interface {
```
Comment (krisis) [inline pkg/flock/lock.go]: Why do we export this interface?

File: pkg/flock/lock_solaris.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+		Len:    0,
+		Pid:    0,
+	}
+)
+
+func open(path string, flag int, perm os.FileMode) (*os.File, error) {
+	return os.OpenFile(path, flag, perm)
+}
+
+func lock(f *os.File) error {
```
Comment (krisis) [inline pkg/flock/lock_solaris.go]: Plain old F_SETLKW wouldn't conflict with a lock on a different fd on the same name (i.e, open file description) by the same minio server process.

**Review by krisis — COMMENTED** [review]
Open file description (OFD) locks on Linux by fcntl(2) and BSD locks
by flock(2) don't provide the semantics that we
require for locking over shared filesystems like NFS or GlusterFS.
- NFS clients translate flock calls to POSIX byte-range locks (without
  OFD semantics) using the NLM protocol. In this situation, two
  requests holding 'conflicting' locks will be considered as an
  attempt to lock by the same (minio server) process and NOT conflict,
  as we require.
- GlusterFS clients are limited by the fact that FUSE kernel module
  [doesn't have OFD support yet](https://sourceforge.net/p/fuse/mailman/message/35301755/)

File: pkg/flock/lock_test.go (hunk 2 above)
Comment (harshavardhana) [reply pkg/flock/lock_test.go]: We fail with ENOENT.

File: pkg/flock/lock.go (hunk 3 above)
Comment (harshavardhana) [reply pkg/flock/lock.go]: Not needed.. actually.

**Comment by harshavardhana** [comment]
Simple PUT operations are working fine, doing `mc mirror` in parallel is holding the server properly on an NFS mount 

```
mc mirror /usr/bin test1/testbucket/ & mc mirror /usr/bin test2/testbucket/
```

There is a problem in multipart seems to be that tempobj which is being appended in a go-routine has `uploadID` of the multipart session associated with it, it is not a unique file name. This will cause problems regardless since we are doing an AppendFile() we cannot differentiate. 

Will be making changes for multipart to use a separate temp UUID and remember it.

File: cmd/fs-v1-multipart.go [hunk 5]
```diff
@@ -417,12 +417,11 @@ func (fs fsObjects) PutObjectPart(bucket, object, uploadID string, partID int, s
 	}
 
 	go func() {
-		// Append the part in background.
-		errCh := fs.bgAppend.append(fs.storage, bucket, object, uploadID, fsMeta)
 		// Also receive the error so that the appendParts go-routine does not block on send.
 		// But the error received is ignored as fs.PutObjectPart() would have already
 		// returned success to the client.
-		<-errCh
+		// Append the part in background.
```
Comment (krisis) [inline cmd/fs-v1-multipart.go]: nit: The comment needs to change to accomadate the code movement.

File: cmd/namespace-lock.go [hunk 6]
```diff
@@ -137,8 +133,14 @@ func (n *nsLockMap) lock(volume, path string, lockSource, opsID string, readLock
 
 	// Locking here can block.
 	if readLock {
+		if globalFSMutex != nil {
```
Comment (krisis) [inline cmd/namespace-lock.go:137]: The namespace lock should be taken before the shared filesystem lock is attempted. This is the only we can prevent races between two request handlers attempting a lock on the same file from the same minio server process.

File: pkg/lock/lock_nix.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	}
+)
+
+func open(path string, flag int, perm os.FileMode) (*os.File, error) {
+	return os.OpenFile(path, flag, perm)
+}
+
+func lock(f *os.File) error {
+	lock := wrlck
+	lock.Pid = int32(os.Getpid())
```
Comment (krisis) [inline pkg/lock/lock_nix.go]: The pid of minio server process will not change. You can initialize the Pid field at the time of initializing ``wrlck`` and ``rdlck`` variables.

File: pkg/lock/lock_nix.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+func rlock(f *os.File) error {
+	lock := rdlck
+	lock.Pid = int32(os.Getpid())
+	return syscall.FcntlFlock(f.Fd(), syscall.F_SETLKW, &lock)
+}
+
+func runlock(f *os.File) error {
+	lock := rdlck
+	lock.Pid = int32(os.Getpid())
+	if err := syscall.FcntlFlock(f.Fd(), syscall.F_GETLK, &lock); err != nil {
```
Comment (krisis) [inline pkg/lock/lock_nix.go]: What is this GETLK for?

File: pkg/lock/lock_nix.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+	"syscall"
+)
+
+var (
+	wrlck = syscall.Flock_t{
+		Type:   syscall.F_WRLCK,
+		Whence: 0,
+		Start:  0,
+		Len:    0,
+		Pid:    0,
```
Comment (krisis) [inline pkg/lock/lock_nix.go]: The Pid can be initialized here, see comment below.

File: pkg/lock/lock_nix.go (hunk 8 above)
Comment (harshavardhana) [reply pkg/lock/lock_nix.go]: I think the sending an runlock on an runlocked file returns error?

File: cmd/fs-v1-multipart.go (hunk 5 above)
Comment (harshavardhana) [reply cmd/fs-v1-multipart.go]: the code shouldn't be moved was a testing change reverting.

File: cmd/namespace-lock.go (hunk 6 above)
Comment (harshavardhana) [reply cmd/namespace-lock.go:137]: Good catch, let me change it.

File: pkg/lock/lock_nix.go (hunk 7 above)
Comment (harshavardhana) [reply pkg/lock/lock_nix.go]: done

File: pkg/lock/lock_nix.go (hunk 9 above)
Comment (harshavardhana) [reply pkg/lock/lock_nix.go]: done

**Comment by harshavardhana** [comment]
We decided to do a completely different style of approach here by removing the necessity for `posix.go` at FS layer.  and let FS layer directly talk to backend.  There is a reason behind that - it provides a way for us to hold locks transparently for each i/o - current semantics of `fcntl` will work for us. 

It also solves the problem of running multiple minio servers on same node with same backend but using different access keys. Will explain more on this will call.

**Comment by harshavardhana** [comment]
Closing, superseded by #3397

**Pull request closed without merging**